core/node/events/stream_ephemeral.go

```diff
@@ -0,0 +1,182 @@
+		}
+	}()
+}
+
+// normalizeEphemeralStream normalizes the ephemeral stream.
+// Loads the missing miniblocks from the sticky peers and writes them to the storage.
+// Seals the stream if it is ephemeral and all miniblocks are loaded.
+func (s *streamCacheImpl) normalizeEphemeralStream(
+	ctx context.Context,
+	stream *streamImpl,
+	lastMiniblockNum int64,
+	isSealed bool,
+) error {
+	if !isSealed {
+		// Stream is not sealed, no need to normalize it yet.
+		return nil
+	}
+
+	missingMbs := make([]int64, 0, lastMiniblockNum+1)
+
+	// Check if the given stream is already sealed, if so, ignore the event.
+	ephemeral, err := s.params.Storage.IsStreamEphemeral(ctx, stream.streamId)
+	if err != nil {
+		if !IsRiverErrorCode(err, Err_NOT_FOUND) {
+			return err
+		}
+
+		// Stream does not exist in the storage - the entire stream is missing.
+		for i := int64(0); i <= lastMiniblockNum; i++ {
+			missingMbs = append(missingMbs, int64(i))
+		}
+	} else if !ephemeral {
+		// Stream exists in the storage and sealed already.
+		return nil
+	} else {
+		// Stream exists in the storage, but not sealed yet, i.e. ephemeral.
+
+		// Get existing miniblock numbers.
+		existingMbs, err := s.params.Storage.ReadEphemeralMiniblockNums(ctx, stream.streamId)
+		if err != nil {
+			return err
+		}
+
+		existingMbsMap := make(map[int64]struct{}, len(existingMbs))
+		for _, num := range existingMbs {
+			existingMbsMap[int64(num)] = struct{}{}
+		}
+
+		for num := int64(0); num <= int64(lastMiniblockNum)+1; num++ {
+			if _, exists := existingMbsMap[num]; !exists {
+				missingMbs = append(missingMbs, num)
+			}
+		}
+	}
+
+	// Fill missing miniblocks
+	if len(missingMbs) > 0 {
+		remotes, _ := stream.GetRemotesAndIsLocal()
+		currentStickyPeer := stream.GetStickyPeer()
+		for range len(remotes) {
+			stub, err := s.params.NodeRegistry.GetNodeToNodeClientForAddress(currentStickyPeer)
+			if err != nil {
+				logging.FromCtx(ctx).Errorw("Failed to get node to node client", "err", err, "streamId", stream.streamId)
+				currentStickyPeer = stream.AdvanceStickyPeer(currentStickyPeer)
+				continue
+			}
+
+			resp, err := stub.GetMiniblocksByIds(ctx, connect.NewRequest[GetMiniblocksByIdsRequest](
+				&GetMiniblocksByIdsRequest{
+					StreamId:     stream.streamId[:],
+					MiniblockIds: missingMbs,
+				},
+			))
+			if err != nil {
+				logging.FromCtx(ctx).Errorw("Failed to get miniblocks from sticky peer", "err", err, "streamId", stream.streamId)
+				currentStickyPeer = stream.AdvanceStickyPeer(currentStickyPeer)
+				continue
+			}
+
+			// Start processing miniblocks from the stream.
+			// If the processing breaks in the middle, the rest of missing miniblocks will be fetched from the next sticky peer.
+			for resp.Receive() {
+				mbInfo, err := NewMiniblockInfoFromProto(resp.Msg().GetMiniblock(), NewParsedMiniblockInfoOpts())
+				if err != nil {
+					return err
```

Comment:
In several places an error is returned. Should the node try the next remote if the current sticky peer is returning an error or garbage?

And in these code paths `resp.Close` isn't called.